### ISSUE TITLE
Adding/removing libraries from Java Build Path can go out of sync

### ISSUE THREAD

**Issue by chanseokoh** [GoogleCloudPlatform/google-cloud-eclipse#2688]
Steps to reproduce:

1. Create a new project.
1. Right-click on the project > _Properties_ > _Java Build Path_ > _Libraries_
1. _Add Libraries..._ > _Google Cloud Platform Libraries_ > _Objectify_ > _Finish_
   - While you're still in the _Libraries_ tab, you'll see that the master container has been added and it contains Objectify when expanded.
1. _Add Libraries..._ > _Google Cloud Platform Libraries_ > _Google Cloud Endpoints_ > _Finish_
   - Note that, in the tab, you don't see Endpoints when expanding the container.
1. _OK_ to exit.

Observations:
1. Endpoints is not added.
1. On the other hand, if you right-click on the master container > _Properties_, then Endpoints is checked.

I fear this might be tricky to fix.

**Comment by briandealwis** [comment]
I think this may be because I removed `IClasspathContainerPageExtension2` from `CloudLibrariesPage` in #2553 as it didn't seemed to be necessary.  In this case, although you clicked on _Add Libraries…_, we're modifying the existing entry as we only support one entry.  `ClasspathContainerWizard#performFinish()` calls `CloudLibrariesPage#getSelection()` to return the new entry, but our `getSelection()` returns the edited  classpath entry as the method is documented to return the new or edited entry.  The one advantage of `IClasspathContainerPageExtension2`, which is relevant here, is that it has `getNewContainers()` which is documented to only return the new classpath entries and avoids `getSelection()`.

**Comment by chanseokoh** [comment]
Closely related to this issue, there is a problem in how our library feature works.

When in the _Libraries_ tab of the _Java Build Path_ properties, adding/removing/modifying the master container should not take effect unless the user presses _OK_ or _Apply_. This is true about adding and removing the container entry itself, but unfortunately, selecting and modifying the library list takes effect immediately (i.e., as soon as `collectLibraryFiles()` is called, the library selection list is serialized and persisted).

To demonstrate the problem.
1. Add Objectify to a project. Close any dialog to make the change permanent.
1. _Properties_ > _Java Build Path_ > _Libraries_
1. Double-click on _Google Cloud Platform Libraries_. You'll see Objectify already selected.
1. Select Endpoints > _Finish_ to close our wizard page
1. _Cancel_ to close the _Properties_ dialog

You'll notice Endpoints JARs have been added, even though you canceled your changes.

**Comment by chanseokoh** [comment]
From a quick look, https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2688#issuecomment-353439221 seems tricky to make it right with the current single-master design where we serialize the container info in separate settings files. Clicking the _Finish_ button in our wizard page means the library wizard is closing while calling `performFinish()`. From the perspective of the wizard, everything is done and its world is collapsing. Yet, if we want to fix the problem, we should outlive the destruction of the wizard and know the outside world about if the wizard was called inside the _Properties_ page, which will probably require serious not-so-elegant hacks, and postpone finalization of library changes only after _Apply_ or _OK_ was pressed on the _Properties_ page.

With the single-master container design, the only way I can think of (for now) to fix all the issues, including https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2688#issue-283292814, is to save all the info about the container (the huge JSON we save about JARs and such) in the container entry itself, instead of saving it on the filesystem. That way, if the user reverts changes in _Properties_ by clicking _Cancel_, all the changes will naturally be reverted with the intermediate classpath entry going away. I think theoretically we can save any info into a classpath entry in the form of classpath attributes (using `JavaCore.newClasspathAttribute()`). However, this means a complete overhaul of our current library framework. We need to give some thoughts on this.

**Comment by chanseokoh** [comment]
> ... the only way I can think of ... is to save all the info about the container ... in the container entry itself

Nah, the problem is that when doing _Add Libraries..._, whether doing so on _Properties_ or not, the library wizard is really about adding a new entry rather than updating an existing classpath entry. So you'll end up seeing the "adding a duplicate classpath entry" error. Technically, the error makes sense in that you clicked _Add Libraries..._ rather than _Edit..._ (or opening the _Properties_ of the classpath entry), but in our library flow, the error may look weird.

**Comment by chanseokoh** [comment]
Another problem I noticed, which feels almost like an Eclipse bug, when utilizing classpath attributes is that attributes aren't being updated when updating a classpath in the _Libraries_ tab. On the other hand, they are properly updated, as expected, when ALT+F5 on the master container on the _Project Explorer_. Seems like only the container path be updated reliably. We might be able to squash all library info into the container path, but that seems impractical, if not impossible, and huge pain.

**Comment by chanseokoh** [comment]
> Another problem I noticed, which feels almost like an Eclipse bug

FTR, I think the bug is in (Neon code) `org.eclipse.jdt.internal.ui.wizards.buildpaths.LibrariesWorkbookPage.openContainerSelectionDialog()` (or I could say the `CPListElement` constructor depending on who's responsible).

```java
private CPListElement[] openContainerSelectionDialog(CPListElement existing) {
  ...
      IClasspathEntry created = BuildPathDialogAccess.configureContainerEntry(...);
      if (created != null) {
        CPListElement elem= new CPListElement(null, fCurrJProject, IClasspathEntry.CPE_CONTAINER, created.getPath(), ! created.equals(existingEntry), null, null);
        return new CPListElement[] { elem };
```
Basically, `created` will be an edited entry with updated attributes, but note that no one attempts to read the attributes (via `created.getExtraAttributes()`). `created` is a local variable to be discarded, and attributes are just lost (in practice, what happens is that old attributes will survive and remain there).

**Comment by chanseokoh** [comment]
This is quite tricky to fix, I think. What if we ditch the Eclipse way of adding libraries and rather have our own standalone library add wizard? That may be easier...

**Comment by elharo** [comment]
No custom wizards please. We want to integrate with eclipse, not replace it. And that's way more effort and time than we should be dedicating to this. Have you tried the change @briandealwis suggested above?

**Comment by chanseokoh** [comment]
That did not help, unfortunately.

Even if it did, it cannot fix https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2688#issuecomment-353439221, which is another problem that is different but closely related.

**Comment by chanseokoh** [comment]
We need to discuss how we could tackle both https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2688#issue-283292814 and https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2688#issuecomment-353439221. (Actually, the former already suffers from the latter problem. It is just that pressing _Cancel_ shows the latter problem clearly noticeable, while pressing _OK_ makes it moot.)

These issues are fundamentally due to our unusual scheme of a dynamically changing container (the single master container).

What happens with the steps in https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2688#issue-283292814 is that, the container is resolved via `LibraryClasspathContainerInitializer.initialize()` only when adding Objectify. It is not called when adding Endpoints. The reason is explained in `org.eclipse.jdt.core.ClasspathContainerInitializer.initialize()`:
```
	 * This method is called by the Java model to give the party that defined
	 * this particular kind of classpath container the chance to install
	 * classpath container objects that will be used to convert classpath
	 * container entries into simpler classpath entries. The method is typically
	 * called exactly once for a given Java project and classpath container
	 * entry. This method must not be called by other clients.
```
In other words, it is the Eclipse framework that calls this method only when necessary. On the other hand, our dynamically changing container scheme requires this method be called (to be precise, the container be resolved) whenever the container is changed. I believe we don't have any control over how this Eclipse framework works, so there seems nothing much we can do here.

However, one thing to note is that, at least we can explicitly resolve the container by calling `BuildPath.runContainerResolverJob()`. So, for example, if we schedule this job indefinitely and periodically (which is a very dirty hack that might not really work in practice), we could at least fix or mitigate the original problem. Even so, this won't fix the second problem (https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/2688#issuecomment-353439221) anyway.

Another thing to note is that, all the problem happens only when using our libraries wizard in the _Libraries_ tab. If users do right-click on a project > _Build Path_ > _Add Libraries..._, it is OK.

**Comment by elharo** [comment]
This is not that high a priority. I don't want to burn an indefinite amount of time on it. If theres no obvious quick fix here, our time would likely be better spent on the installer.

**Comment by lak-proddev** [comment]
I believe It's a critical bug and this operation should be in an asynchronous thread. Since the UI is blocking while adding a new library. 
This is my initial impression only. I need to check more details about the bug and code base.

**Comment by lak-proddev** [comment]
while modifying the libraries for Non-Maven project

![image](https://user-images.githubusercontent.com/39690425/59588660-bdff4d00-9105-11e9-86cd-466d15a64bc2.png)

**Comment by briandealwis** [comment]
@lak-proddev This issue is specifically around inconsistencies in using the project _Build Path_ properties to add GCP libraries to the class path.  Sync/async issues should be opened as a new issue.  That said, library resolving and attaching is now async with #3465.  I see a lot of build-path related jobs appear even with a single project, so I don't

**Comment by chanseokoh** [comment]
@lak-proddev #3465 is merged recently (not released yet). Is the screenshot on recent master (HEAD)? That is, is it after #3465 or not?

**Comment by lak-proddev** [comment]
Yes, It's after   #3465 (commit 00014d662b6240b9d4bb93af71ffbd7614d5492c). The change  included in my workspace.

**Comment by briandealwis** [comment]
@lak-proddev I meant to add: if you were inquiring about the number of build path jobs, we attempt to download the source jars for each of the resolved dependencies in a separate job.  So it's not unexpected to see a lot of jobs fired up.

I wonder if we need to do this as m2e provides a source connector that examines the jars and, if a maven artifact, will automatically download and configure the source.  I don't remember when this was introduced though.

**Comment by lak-proddev** [comment]
> @lak-proddev I meant to add: if you were inquiring about the number of build path jobs ......
Yes, I just want to let u know. Since multiple jobs running parallel. Nothing more than that.

**Comment by SINGHPRT** [comment]
Adding  to following section in  .classpath file made it appear in Build path.

<classpathentry kind="con" path="com.google.cloud.tools.eclipse.appengine.libraries/master-container">
		<attributes>
			<attribute name="maven.pomderived" value="true"/>
			<attribute name="org.eclipse.jst.component.dependency" value="/WEB-INF/lib"/>
		</attributes>
	</classpathentry>

**Comment by JoeWang1127** [comment]
close as not planned
